nheko: add identicons support

unclechu opened the pull request:
###### Motivation for this change

Nheko has support for identicons. There is a checkbox in the settings of nheko but it wasn’t clickable because the dependency (“qt-jdenticon” plugin) was missing.

This change adds “qt-jdenticon” as a new package (Qt plugin) and as a dependency for Nheko. This package is maintained by Nheko team.

![2021-12-13 00-59-48](https://user-images.githubusercontent.com/799353/145733431-1af540a5-cb69-476d-a1ee-b1187983abc1.png)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

unclechu commented:
Context reference: https://github.com/NixOS/nixpkgs/pull/150437#issuecomment-991965132

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28044,6 +28044,8 @@ with pkgs;
 
   qxw = callPackage ../applications/editors/qxw {};
 
+  qt-jdenticon = libsForQt5.callPackage ../development/libraries/qt-jdenticon {};
```
This is a library hence it should be inside `qt5-packages.nix`.

unclechu replied:
Now it’s added to `qt6-packages.nix` instead of `all-packages.nix`

doronbehar commented on pkgs/development/libraries/qt-jdenticon/default.nix:
```diff
@@ -0,0 +1,32 @@
+  nativeBuildInputs = [ qmake ];
+  buildInputs = [ qtbase ];
+
+  dontWrapQtApps = true;
```
Why do you need this?

unclechu replied:
I quote a paragraph from `doc/languages-frameworks/qt.section.md` file:

``` html
    <para>
      All Qt packages must include <literal>wrapQtAppsHook</literal> in
      <literal>nativeBuildInputs</literal>, or you must explicitly set
      <literal>dontWrapQtApps</literal>.
    </para>
```

Before I added this option the build of the package was producing an empty directory. If I’m not mistaken just by adding this option it made everything work as expected.

doronbehar replied:
> I quote a paragraph from doc/languages-frameworks/qt.section.md file:

Fair enough, I forgot :).

doronbehar reviewed (changes requested):
And please change the git history into two commits:

```
libsForQt5.qt-jdenticon: init at 0.2.1
```

And:

```
nheko: Add support for `jdenticon`
```

Or alike.

unclechu commented:
@doronbehar 

> qt-jdenticon is meant for qt5, and nheko is built with qt6. Closing.

It works for me on Nheko build for Qt6? Here is a demonstration:

https://github.com/unclechu/nixos-config/blob/1fcdb7e65cb21e523c81adc5c6d5cbea947a6cd5/apps/nheko/default.nix#L28-L59

https://github.com/unclechu/nixos-config/blob/1fcdb7e65cb21e523c81adc5c6d5cbea947a6cd5/apps/nheko/default.nix#L73-L76

It works both for Qt5 and Qt6 just fine.

doronbehar commented:
Hmm OK, then please add it to both `qt5-packages.nix` and `qt6-packages.nix`, as I asked you already 3 years ago :) fix the merge conflicts, and we'll see. Also, remember to separate the `qt-jdenticon: init` commit to a separate commit.

unclechu committed:
qt6Packages.qt-jdenticon: init at 0.3.0

unclechu committed:
nheko: Add support for `jdenticon`

unclechu commented:
@doronbehar I fixed everything, please have another look.

I think I was wrong about `qt-jdenticon` was working both for Qt5 and Qt6. `0.2.*` was made for Qt5 but I was actually using `0.3.0` that is made to work with Qt6, with newer Nheko. So I didn’t add the package to `qt5-packages.nix`, I assume it won’t work judging by CMake configuration file which only mentions Qt6. But I added the package to `qt6-packages.nix` (removed from `all-packages.nix`).

- So now it’s 0.3.0 version instead of 0.2.1
- I made separate commits for adding `qt-jdenticon` package and using the package for Nheko
- Applied new `nixfmt`
- Tested that the library package builds like this: `nix-build -E 'with import ./. {}; qt6Packages.qt-jdenticon'`
- And built Nheko like this: `nix-build -E 'with import ./. { config.permittedInsecurePackages=["olm-3.2.16"]; }; nheko'` and tested that jdenticons are rendered properly: ![2024-12-22 07-06-14](https://github.com/user-attachments/assets/d54b6c6d-86ee-4ccd-a863-b46d964e3025)

doronbehar commented:
I added a few minor modifications which were hard to explain in a review :), and forced pushed. I'm sure you won't mind.

doronbehar commented:
Oh I'm sorry for the mass ping everyone! I have no idea how could this rebase go wrong!
